Documentation for RP-Initiated Logout refactoring

mposolda opened the pull request:
Closes #1442

Codebase PR: https://github.com/keycloak/keycloak/pull/10887

mposolda committed:
Documentation for RP-Initiated Logout refactoring
Closes #1442

andymunro commented on release_notes/topics/18_0_0.adoc:
```diff
@@ -4,6 +4,17 @@
 
+== OpenID Connect Logout Improvements
+
+Few fixes and improvements were made to make sure that {project_name} is now fully compliant with all the OpenID Connect logout specifications:
```
Change "Few" to "Some"

andymunro commented on release_notes/topics/18_0_0.adoc:
```diff
@@ -4,6 +4,17 @@
+* OpenID Connect Back-Channel Logout 1.0
+* OpenID Connect Session Management 1.0
+
+For more details, take a look at link:{adminguide_link}#_oidc-logout[{adminguide_name}].
```
We are standardizing on  "For more details, see.."

mposolda replied:
Ok, so I am changing it also for 3 other places in this file 18_0_0.adoc. Thanks

andymunro commented on securing_apps/topics/oidc/java/logout.adoc:
```diff
@@ -3,12 +3,17 @@
 You can log out of a web application in multiple ways.
 For Jakarta EE servlet containers, you can call `HttpServletRequest.logout()`. For other browser applications, you can redirect the browser to
-`\http://auth-server{kc_realms_path}/{realm-name}/protocol/openid-connect/logout?redirect_uri=encodedRedirectUri`, which logs you out if you have an SSO session with your browser.
+`\http://auth-server{kc_realms_path}/{realm-name}/protocol/openid-connect/logout`, which logs the user out if he has an SSO session with his browser. The actual logout is done once user
```
Change "he" to "that user"
add "the" after "once"

andymunro commented on securing_apps/topics/oidc/java/logout.adoc:
```diff
@@ -3,12 +3,17 @@
 For Jakarta EE servlet containers, you can call `HttpServletRequest.logout()`. For other browser applications, you can redirect the browser to
-`\http://auth-server{kc_realms_path}/{realm-name}/protocol/openid-connect/logout?redirect_uri=encodedRedirectUri`, which logs you out if you have an SSO session with your browser.
+`\http://auth-server{kc_realms_path}/{realm-name}/protocol/openid-connect/logout`, which logs the user out if he has an SSO session with his browser. The actual logout is done once user
+confirms the logout. You can optionally include parameters like `id_token_hint`, `post_logout_redirect_uri` and others as described in the
```
Generally replace "like" with "such as"

andymunro commented on securing_apps/topics/oidc/java/logout.adoc:
```diff
@@ -3,12 +3,17 @@
-`\http://auth-server{kc_realms_path}/{realm-name}/protocol/openid-connect/logout?redirect_uri=encodedRedirectUri`, which logs you out if you have an SSO session with your browser.
+`\http://auth-server{kc_realms_path}/{realm-name}/protocol/openid-connect/logout`, which logs the user out if he has an SSO session with his browser. The actual logout is done once user
+confirms the logout. You can optionally include parameters like `id_token_hint`, `post_logout_redirect_uri` and others as described in the
+https://openid.net/specs/openid-connect-rpinitiated-1_0.html[OpenID Connect RP-Initiated Logout], which can cause that logout does not need to be explicitly confirmed
```
Change `,` which can cause" to ". As a result,"

andymunro commented on securing_apps/topics/oidc/java/logout.adoc:
```diff
@@ -3,12 +3,17 @@
+by the user.
+
+If you want to avoid logging out of an external identity provider as part of the logout process, you can supply the parameter `$$initiating_idp$$`, with the value being
+the identity (alias) of the identity provider in question. This is useful when the logout endpoint is invoked as part of single logout initiated by the external identity provider.
```
This parameter

andymunro commented on server_admin/topics/sso-protocols/con-oidc-auth-flows.adoc:
```diff
@@ -312,9 +312,10 @@ User Resolver Provider is provided as SPI provider so that users of {project_nam
 ==== OIDC Logout
 
-OIDC has three different specifications relevant to logout mechanisms, all of these are currently in draft status:
+OIDC has four different specifications relevant to logout mechanisms, all of these are currently in draft status:
```
OIDC has four specifications relevant to logout mechanisms. These parameters are in draft status:

mposolda replied:
I believe you meant "These specifications..." in the second sentence?  I think that "These parameters..." does not makes much sense here.

andymunro commented on server_admin/topics/sso-protocols/con-oidc-auth-flows.adoc:
```diff
@@ -325,15 +326,21 @@ Again since all of this is described in the OIDC specification we will only give
+===== RP-Initiated Logout
 
 This is also a browser-based logout where the logout starts by redirecting the user to a specific endpoint at {project_name}.
+This redirect usually happens when user press `Log Out` link on the page of some application, which previously used {project_name} to authenticate the user.
```
when the user clicks the

andymunro commented on server_admin/topics/sso-protocols/con-oidc-auth-flows.adoc:
```diff
@@ -325,15 +326,21 @@ Again since all of this is described in the OIDC specification we will only give
 Once the user is redirected to the logout endpoint, {project_name} is going to send logout requests to
 clients to let them to invalidate their local user sessions, and potentially redirect the user to some URL
-once the logout process is finished.
+once the logout process is finished. User might be optionally requested to confirm the logout in case that `id_token_hint` parameter was not used
```
The user...

andymunro commented on server_admin/topics/sso-protocols/con-oidc-auth-flows.adoc:
```diff
@@ -325,15 +326,21 @@ Again since all of this is described in the OIDC specification we will only give
+or the client referenced in the ID Token from `id_token_hint` has `Consent Required` enabled.
+
+Depending on the client configuration, logout requests can be sent to clients through the front-channel or through the back-channel. For the frontend browser clients, which rely on the
+Session Management described above, {project_name} does not need to send any logout requests to them because these clients are automatically able to detect that SSO session
```
change "described above" to "described in the previous section"
change "to them because these clients are automatically able to detect that SSO session"
to "to them;  these clients automatically detect that SSO session"

andymunro commented on upgrading/topics/keycloak/changes-18_0_0.adoc:
```diff
@@ -18,3 +18,41 @@ the migration. Consider these possible actions:
+
+= OpenID Connect Logout
+
+Previous versions of {project_name} had support for automatic logout of the user and redirecting to the application by opening logout endpoint URL like
```
change "support for automatic logout"
  to active voice
"supported automatic logout"  
change "like" to "such as"

andymunro commented on upgrading/topics/keycloak/changes-18_0_0.adoc:
```diff
@@ -18,3 +18,41 @@ the migration. Consider these possible actions:
+= OpenID Connect Logout
+
+Previous versions of {project_name} had support for automatic logout of the user and redirecting to the application by opening logout endpoint URL like
+`https://auth-server/auth/realms/{realm-name}/protocol/openid-connect/logout?redirect_uri=encodedRedirectUri`. This was easy to use, but had potentially negative impact
```
change "This was easy to use, but
to
"While that implementation was easy to use, it"

andymunro commented on upgrading/topics/keycloak/changes-18_0_0.adoc:
```diff
@@ -18,3 +18,41 @@ the migration. Consider these possible actions:
+
+Previous versions of {project_name} had support for automatic logout of the user and redirecting to the application by opening logout endpoint URL like
+`https://auth-server/auth/realms/{realm-name}/protocol/openid-connect/logout?redirect_uri=encodedRedirectUri`. This was easy to use, but had potentially negative impact
+on performance and security. The new version has proper support for logout based on the OpenID Connect RP-Initiated Logout specification. The parameter `redirect_uri` is not
```
has better support

andymunro commented on upgrading/topics/keycloak/changes-18_0_0.adoc:
```diff
@@ -18,3 +18,41 @@ the migration. Consider these possible actions:
+Previous versions of {project_name} had support for automatic logout of the user and redirecting to the application by opening logout endpoint URL like
+`https://auth-server/auth/realms/{realm-name}/protocol/openid-connect/logout?redirect_uri=encodedRedirectUri`. This was easy to use, but had potentially negative impact
+on performance and security. The new version has proper support for logout based on the OpenID Connect RP-Initiated Logout specification. The parameter `redirect_uri` is not
+supported anymore and user needs to confirm the logout. It is possible to omit the confirmation and do automatic redirect to the application when you include parameter `post_logout_redirect_uri`
```
Change "The parameter `redirect_uri` is not supported anymore and user needs to confirm the logout."
to 
"The parameter `redirect_uri` is no longer supported; also, in the new version, the user needs to confirm the logout."

andymunro commented on upgrading/topics/keycloak/changes-18_0_0.adoc:
```diff
@@ -18,3 +18,41 @@ the migration. Consider these possible actions:
+supported anymore and user needs to confirm the logout. It is possible to omit the confirmation and do automatic redirect to the application when you include parameter `post_logout_redirect_uri`
+together with the parameter `id_token_hint` with the ID Token used for login.
+
+The existing deployments are affected as described below:
```
The existing deployments are affected in the following ways:

andymunro commented on upgrading/topics/keycloak/changes-18_0_0.adoc:
```diff
@@ -18,3 +18,41 @@ the migration. Consider these possible actions:
+
+The existing deployments are affected as described below:
+
+- If your application directly uses links to logout endpoint with the `redirect_uri` parameter, you may be required to change this as described above.
```
use "log out" when it is a verb -  "log out an endpoint"

mposolda replied:
This is actually not the verb. The "logout endpoint" is the name of the endpoint. It is the REST endpoint, which is used for logout user and we call it "Logout endpoint". So for now, I keep this part.

For more illustration, see for example this part of the existing documentation https://www.keycloak.org/docs/latest/securing_apps/index.html#logout and the last sentence: This is useful when the logout endpoint is invoked as part of single logout initiated by the external identity provider.

andymunro replied:
Okay, thanks for the clarification, @mposolda.

andymunro commented on upgrading/topics/keycloak/changes-18_0_0.adoc:
```diff
@@ -18,3 +18,41 @@ the migration. Consider these possible actions:
+The existing deployments are affected as described below:
+
+- If your application directly uses links to logout endpoint with the `redirect_uri` parameter, you may be required to change this as described above.
+  It is recommeded to either remove `redirect_uri` parameter entirely or replace it with `id_token_hint` and `post_logout_redirect_uri` parameters.
```
Consider either removing the `redirect_uri` parameter entirely or replacing it with the `id_token_hint` and `post_logout_redirect_uri` parameters.

andymunro commented on upgrading/topics/keycloak/changes-18_0_0.adoc:
```diff
@@ -18,3 +18,41 @@ the migration. Consider these possible actions:
+
+- If your application directly uses links to logout endpoint with the `redirect_uri` parameter, you may be required to change this as described above.
+  It is recommeded to either remove `redirect_uri` parameter entirely or replace it with `id_token_hint` and `post_logout_redirect_uri` parameters.
+- If you use java adapters and your application does logout by call `httpServletRequest.logout()`, you are not affected because this call uses the backchannel variant of the logout endpoint
```
I think these occurrences of "logout" are a noun and an adjective, so "logout" is okay here.

mposolda replied:
Yes, exactly.

andymunro commented on upgrading/topics/keycloak/changes-18_0_0.adoc:
```diff
@@ -18,3 +18,41 @@ the migration. Consider these possible actions:
+  It is recommeded to either remove `redirect_uri` parameter entirely or replace it with `id_token_hint` and `post_logout_redirect_uri` parameters.
+- If you use java adapters and your application does logout by call `httpServletRequest.logout()`, you are not affected because this call uses the backchannel variant of the logout endpoint
+  and that one was not changed.
+- If you use latest javascript adapter, you are also not affected. However if your application uses older version of javascript adapter, you are affected as this
```
"the" latest
"an older version of the JavaScript adapter"

andymunro commented on upgrading/topics/keycloak/changes-18_0_0.adoc:
```diff
@@ -18,3 +18,41 @@ the migration. Consider these possible actions:
+- If you use java adapters and your application does logout by call `httpServletRequest.logout()`, you are not affected because this call uses the backchannel variant of the logout endpoint
+  and that one was not changed.
+- If you use latest javascript adapter, you are also not affected. However if your application uses older version of javascript adapter, you are affected as this
+  adapter uses the variant of the logout endpoint with the deprecated `redirect_uri` parameter. In this case, you may need to upgrade to the latest version of the javascript adapter.
```
JavaScript

andymunro commented on upgrading/topics/keycloak/changes-18_0_0.adoc:
```diff
@@ -18,3 +18,41 @@ the migration. Consider these possible actions:
+- If you use latest javascript adapter, you are also not affected. However if your application uses older version of javascript adapter, you are affected as this
+  adapter uses the variant of the logout endpoint with the deprecated `redirect_uri` parameter. In this case, you may need to upgrade to the latest version of the javascript adapter.
+
+There is backwards compatibility option, which allows that your application can still use the old format of the `redirect_uri` parameter. It is possible to enable it
```
There is a backwards compatibility option, which allows your application to still use the old format of the `redirect_uri` parameter. 

ifeval::["{kc_dist}" == "quarkus"]
You can enable this parameter when you start the server by entering the following command:
bin/kc.[sh|bat]  --spi-login-protocol-openid-connect-legacy-logout-redirect-uri=true start
endif::[]

ifeval::["{kc_dist}" == "wildfly"]
You can enable this parameter by including the following configuration in the `standalone-*.xml` file:

[source,bash,subs=+attributes]
```
<spi name="login-protocol">
    <provider name="openid-connect" enabled="true">
        <properties>
            <property name="legacy-logout-redirect-uri" value="true"/>
        </properties>
    </provider>
</spi>
```
endif::[]

With this configuration, you can still use the format with the `redirect_uri` parameter. Note the confirmation screen will be needed if the `id_token_hint` is omitted.

andymunro reviewed (changes requested):
A few suggestions, @mposolda. Looks very complete

The suggestions at the end would hide references to Quarkus content in WildFly documentation if we have a WildFly version of the Upgrading Guide.

I am using the same "ifeval" statements that @stianst is using.

mposolda commented:
@andymunro Thanks! I've addressed almost all your suggestions and I resolved relevant comments, which I resolved. I left only one, where I left unresolved and added comment to this one.

andymunro reviewed (dismissed):
Looks great, @mposolda.  You may still want to check if there will be a separate upgrading guide for the WildFly distribution.  The current upgrading guide, and the changes I suggested in the PR, shows content only for Quarkus.

stianst commented on upgrading/topics/keycloak/changes-18_0_0.adoc:
````diff
@@ -18,3 +18,48 @@ the migration. Consider these possible actions:
+You can enable this parameter when you start the server by entering the following command:
+
+```
+bin/kc.[sh|bat] --spi-login-protocol-openid-connect-legacy-logout-redirect-uri=true start
````
```suggestion
bin/kc.[sh|bat] start --spi-login-protocol-openid-connect-legacy-logout-redirect-uri=true
```

mposolda committed:
Added warning about backwards compatibility switch removal in the future

mposolda commented:
@stianst @andymunro Updated this PR with one small warning in the upgrading guide, that switch for backwards compatibility will be removed in the future. I've used separate commit just so you can easier review the changes, but I plan to squash the commits before merging this PR.
